sdk/storage/inc/common/concurrent_transfer.hpp

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <atomic>
+#include <cstdlib>
+#include <functional>
+#include <future>
+
+namespace Azure { namespace Storage { namespace Details {
+
+  inline void ConcurrentTransfer(
+      int64_t offset,
+      int64_t length,
+      int64_t chunkSize,
+      int concurrency,
+      // offset, length, chunk id, number of chunks
+      std::function<void(int64_t, int64_t, int64_t, int64_t)> transferFunc)
+  {
+    std::atomic<int> numWorkingThreads{concurrency};
+    std::atomic<int> nextChunkId{0};
+    std::atomic<bool> failed{false};
+
+    const auto numChunks = (length + chunkSize - 1) / chunkSize;
+
+    auto threadFunc = [&]() {
+      while (true)
+      {
+        int chunkId = nextChunkId.fetch_add(1);
+        if (chunkId >= numChunks || failed)
+        {
+          break;
+        }
+        int64_t chunkOffset = offset + chunkSize * chunkId;
+        int64_t chunkLength = std::min(length - chunkSize * chunkId, chunkSize);
+        try
+        {
+          transferFunc(chunkOffset, chunkLength, chunkId, numChunks);
```

Comment:
Put block will be one of the transferFunc, but the chunk ID cannot be passed out, meaning that the caller does not know the chunk ID being passed into the Put block operation if chunk ID is not a number starting from 0.
Adding a chunk ID prefix parameter to the `ConcurrentTransfer` and make sure chunkId is a string instead of an int will resolve this issue.

Reply:
Hmm, saw your solution, ignore this comment then.